src/dtx/dtx_common.c

```diff
@@ -1697,50 +1700,53 @@ struct dtx_ult_arg {
 static void
 dtx_leader_exec_ops_ult(void *arg)
 {
-	struct dtx_ult_arg	  *ult_arg = arg;
-	struct dtx_leader_handle  *dlh = ult_arg->dlh;
-	ABT_future		  future = dlh->dlh_future;
-	uint32_t		  i;
-	int			  rc = 0;
+	struct dtx_ult_arg		*ult_arg = arg;
+	struct dtx_leader_handle	*dlh = ult_arg->dlh;
+	struct dtx_sub_status		*sub;
+	ABT_future			 future = dlh->dlh_future;
+	uint32_t			 i;
+	int				 rc = 0;
 
 	D_ASSERT(future != ABT_FUTURE_NULL);
 	for (i = 0; i < dlh->dlh_sub_cnt; i++) {
-		struct dtx_sub_status *sub = &dlh->dlh_subs[i];
-
+		sub = &dlh->dlh_subs[i];
 		sub->dss_result = 0;
+		sub->dss_comp = 0;
 
 		if (sub->dss_tgt.st_rank == DAOS_TGT_IGNORE ||
 		    (i == daos_fail_value_get() &&
 		     DAOS_FAIL_CHECK(DAOS_DTX_SKIP_PREPARE))) {
-			int ret;
-
-			ret = ABT_future_set(future, dlh);
-			D_ASSERTF(ret == ABT_SUCCESS,
-				  "ABT_future_set failed %d.\n", ret);
+			sub->dss_comp = 1;
+			rc = ABT_future_set(future, dlh);
+			D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_set failed (1) %d.\n", rc);
 			continue;
 		}
 
-		rc = ult_arg->func(dlh, ult_arg->func_arg, i, dtx_sub_comp_cb);
-		if (rc) {
-			sub->dss_result = rc;
-			break;
-		}
-
 		/* Yield to avoid holding CPU for too long time. */
 		if (i >= DTX_RPC_YIELD_THD)
 			ABT_thread_yield();
+
+		rc = ult_arg->func(dlh, ult_arg->func_arg, i, dtx_sub_comp_cb);
+		if (rc != 0) {
+			if (sub->dss_comp == 0)
+				dtx_sub_comp_cb(dlh, i, rc);
+			break;
+		}
 	}
 
 	if (rc != 0) {
 		for (i++; i < dlh->dlh_sub_cnt; i++) {
-			int ret;
-
-			ret = ABT_future_set(future, dlh);
-			D_ASSERTF(ret == ABT_SUCCESS,
-				  "ABT_future_set failed %d.\n", ret);
+			sub = &dlh->dlh_subs[i];
+			sub->dss_comp = 1;
+			rc = ABT_future_set(future, dlh);
+			D_ASSERTF(rc == ABT_SUCCESS, "ABT_future_set failed (2) %d.\n", rc);
 		}
 	}
 
+	/* To indicate that the IO forward ULT itself done. */
+	rc = ABT_future_set(future, dlh);
```

Comment:
A simpler way could be calling dss_ult_execute() instead of dss_ult_create() in dtx_leader_exec_ops(). Actually I think waiting for all RPCs done here (in forwarding ULT itself) looks more straightforward.

Reply:
It is the main IO handler ULT to wait the RPCs instead of the IO forward ULT.

Reply:
right, we still want to create all ULTs in a row